### PR TITLE
[exporter/otlphttp] added support for configurable telemetry encoding

### PR DESCRIPTION
**Pull request by tvaintrob**
**Description:**
This PR adds support for encoding configuration in the `otlphttp` exporter.

**Link to tracking Issue:** 
#6945

**Testing:** 
Updated existing tests, and added relevant tests 

**Documentation:** 
Updated the `otlphttp` docs to include the new configuration option.

### CODE REVIEW

File: exporter/otlphttpexporter/otlp_test.go
Comment (songy23) [inline exporter/otlphttpexporter/otlp_test.go:652]: Can you add tests on malformed data? Same for metrics and logs

**Review by songy23 — COMMENTED** [review]
Thanks @tvaintrob please also add a changelog

**Comment by bogdandrutu** [comment]
What is the use-case for this? I remember that we intentionally did not want to offer this, because it is just a slower version of the proto and you don't have the dependency issue like you have on the client side (web) to  avoid depending on protobuf.

**Comment by tvaintrob** [comment]
@bogdandrutu in my case we have a WAF that we want to send messages through but it needs to do a deep validation of the messages being passed, unfortunately it doesn't support reading binary data so we need to send the messages using JSON

File: exporter/otlphttpexporter/otlp_test.go
Comment (tvaintrob) [reply exporter/otlphttpexporter/otlp_test.go:652]: what do you mean by malformed data? invalid `encoding` field?

**Comment by tvaintrob** [comment]
@songy23 fixed all comments except for the additional tests

File: exporter/otlphttpexporter/otlp_test.go
Comment (songy23) [reply exporter/otlphttpexporter/otlp_test.go:652]: I mean invalid proto or json text data, i.e. when err is not nil.

**Review by songy23 — COMMENTED** [review]
Please fix the failed tests https://github.com/open-telemetry/opentelemetry-collector/actions/runs/7513270300/job/20574065081?pr=9276

**Comment by songy23** [comment]
We can mention in the documentation that json should only be used in endpoints that can only accept json, but discouraged in other use cases, e.g. collector-to-collector case. Does that sound good? @bogdandrutu

**Comment by TylerHelmuth** [comment]
I agree with supporting `http/json`. Although it isn't performant, it feels correct to implement all parts of the OTLP specification for the exporter.

File: exporter/otlphttpexporter/otlp_test.go
Comment (songy23) [reply exporter/otlphttpexporter/otlp_test.go:652]: Chatted offline, with the current setup it's difficult to generate data that cannot be parsed, so mark this as resolved.

File: exporter/otlphttpexporter/config_test.go
Comment (songy23) [inline exporter/otlphttpexporter/config_test.go:54]: Why is this line needed?

File: exporter/otlphttpexporter/factory_test.go
Comment (songy23) [inline exporter/otlphttpexporter/factory_test.go:39]: Need to update `factory.CreateDefaultConfig()` to set default encoding to EncodingProto

Comment (tvaintrob) [reply exporter/otlphttpexporter/factory_test.go:39]: This broke the tests from contrib, and because the default handler is already proto it seemed fine to leave empty

File: exporter/otlphttpexporter/factory.go
Comment (tvaintrob) [reply exporter/otlphttpexporter/factory.go]: How do I go about doing it? The contrib repo doesn't have a reference to the updated config struct

Comment (songy23) [reply exporter/otlphttpexporter/factory.go]: OK, let's leave a TODO here and update the contrib tests when the PR is merged and the core dependency is updated in contrib. Default should be made explicit IMO.

Comment (songy23) [reply exporter/otlphttpexporter/factory.go]: Discussed offline, there's a circular dependency between the otlphttpexporter config and a test in f5cloudexporter https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/30703. So for now the default cannot be changed.

**Review by songy23 — DISMISSED** [review]
LGTM

cc @open-telemetry/collector-approvers PTAL

File: exporter/otlphttpexporter/config.go
Comment (TylerHelmuth) [inline exporter/otlphttpexporter/config.go:16]: Should we implement `UnmarshalText` to force this type to be either `proto` or `json` when used in a config?

File: exporter/otlphttpexporter/config.go
Comment (TylerHelmuth) [inline exporter/otlphttpexporter/config.go:39]: I think we should have some validation, either as part of `Validate` or `UnmarshalText`, that enforces only `proto` or `json`.

File: exporter/otlphttpexporter/factory_test.go
Comment (TylerHelmuth) [reply exporter/otlphttpexporter/factory_test.go:39]: I agree that `factory.CreateDefaultConfig()` needs updated to set `Encoding: EncodingProto`.  What broke in contrib?

File: exporter/otlphttpexporter/otlp.go
Comment (TylerHelmuth) [inline exporter/otlphttpexporter/otlp.go:99]: Since we have explicit cases for the encodings we handle I think this should error instead of fallback to proto.

File: exporter/otlphttpexporter/otlp.go
Comment (TylerHelmuth) [inline exporter/otlphttpexporter/otlp.go:120]: Since we have explicit cases for the encodings we handle I think this should error instead of fallback to proto.

File: exporter/otlphttpexporter/otlp.go
Comment (TylerHelmuth) [inline exporter/otlphttpexporter/otlp.go:140]: Since we have explicit cases for the encodings we handle I think this should error instead of fallback to proto.

File: exporter/otlphttpexporter/otlp.go
Comment (TylerHelmuth) [inline exporter/otlphttpexporter/otlp.go:163]: Since we have explicit cases for the encodings we handle I think this should error instead of fallback to proto.

File: exporter/otlphttpexporter/otlp.go
Comment (TylerHelmuth) [inline exporter/otlphttpexporter/otlp.go:318]: I think this could also be a switch statement that errors if one of the expected content types is not received.

File: exporter/otlphttpexporter/otlp.go
Comment (TylerHelmuth) [inline exporter/otlphttpexporter/otlp.go:343]: I think this could also be a switch statement that errors if one of the expected content types is not received.

File: exporter/otlphttpexporter/otlp.go
Comment (TylerHelmuth) [inline exporter/otlphttpexporter/otlp.go:368]: I think this could also be a switch statement that errors if one of the expected content types is not received.

File: exporter/otlphttpexporter/factory_test.go
Comment (tvaintrob) [reply exporter/otlphttpexporter/factory_test.go:39]: The `f5cloudexporter` broke because one of the tests explicitly creates an instance of the config as part of a test, the new field caused it to mismatch with the asserted one

See https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/30703

File: exporter/otlphttpexporter/otlp.go
Comment (tvaintrob) [reply exporter/otlphttpexporter/otlp.go:99]: Since there is an issue with adding the default value explicitly (breaking contrib) I cant change this, unless breaking contrib is acceptable (I can create a small fix in the `f5cloudexporter` once this change is merged)

File: exporter/otlphttpexporter/otlp.go
Comment (tvaintrob) [reply exporter/otlphttpexporter/otlp.go:120]: Since there is an issue with adding the default value explicitly (breaking contrib) I cant change this, unless breaking contrib is acceptable (I can create a small fix in the `f5cloudexporter` once this change is merged)

File: exporter/otlphttpexporter/otlp.go
Comment (tvaintrob) [reply exporter/otlphttpexporter/otlp.go:140]: Since there is an issue with adding the default value explicitly (breaking contrib) I cant change this, unless breaking contrib is acceptable (I can create a small fix in the `f5cloudexporter` once this change is merged)

File: exporter/otlphttpexporter/otlp.go
Comment (tvaintrob) [reply exporter/otlphttpexporter/otlp.go:163]: Since there is an issue with adding the default value explicitly (breaking contrib) I cant change this, unless breaking contrib is acceptable (I can create a small fix in the `f5cloudexporter` once this change is merged)

File: exporter/otlphttpexporter/otlp.go
Comment (tvaintrob) [reply exporter/otlphttpexporter/otlp.go:318]: i changed to a switch statement but left previous behaviour, before this PR if the content-type was not proto a `nil` was returned

File: exporter/otlphttpexporter/otlp.go
Comment (tvaintrob) [reply exporter/otlphttpexporter/otlp.go:343]: i changed to a switch statement but left previous behaviour, before this PR if the content-type was not proto a `nil` was returned

File: exporter/otlphttpexporter/otlp.go
Comment (tvaintrob) [reply exporter/otlphttpexporter/otlp.go:368]: i changed to a switch statement but left previous behaviour, before this PR if the content-type was not proto a `nil` was returned

**Comment by tvaintrob** [comment]
@TylerHelmuth updated the PR with the requested changes, not all are possible due to the issues with contrib

Thanks

File: exporter/otlphttpexporter/otlp.go
Comment (TylerHelmuth) [reply exporter/otlphttpexporter/otlp.go:99]: Can you share more about what breaks in contrib? Can we go fix it now and then merge this PR?

File: exporter/otlphttpexporter/otlp.go
Comment (TylerHelmuth) [inline exporter/otlphttpexporter/otlp.go:317]: I agree that returning nil was the previous default behavior, but I think that was incorrect.  I am ok taking care of that in a separate PR tho, as long as we create an issue to track it.

File: exporter/otlphttpexporter/config.go
Comment (TylerHelmuth) [inline exporter/otlphttpexporter/config.go:31]: I believe we can get by without adding `MarshalText`, reducing our public API.

**Review by TylerHelmuth — COMMENTED** [review]
Thanks for opening https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/30703.  I'd say we fix that before merging this PR.

**Comment by songy23** [comment]
> Thanks for opening https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/30703. I'd say we fix that before merging this PR.

Turns out to be straightforward to fix it, I have https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30759

File: exporter/otlphttpexporter/factory_test.go
Comment (songy23) [reply exporter/otlphttpexporter/factory_test.go:39]: The test in f5cloudexporter is fixed, you should be able to update `factory.CreateDefaultConfig()` now

**Comment by tvaintrob** [comment]
Thanks for the update @songy23 ill update this one to include the default encoding

File: exporter/otlphttpexporter/config.go
Comment (songy23) [inline exporter/otlphttpexporter/config.go:27]:
```suggestion
// UnmarshalText unmarshalls text to an EncodingType.
```

File: exporter/otlphttpexporter/otlp.go
Comment (songy23) [reply exporter/otlphttpexporter/otlp.go:317]: I think this should return an error with unsupported content type now?

File: exporter/otlphttpexporter/otlp.go
Comment (songy23) [inline exporter/otlphttpexporter/otlp.go:341]: Same here

File: exporter/otlphttpexporter/otlp.go
Comment (songy23) [inline exporter/otlphttpexporter/otlp.go:365]: Same here

**Review by songy23 — COMMENTED** [review]
Minor comments, LGTM otherwise

File: exporter/otlphttpexporter/otlp.go
Comment (TylerHelmuth) [reply exporter/otlphttpexporter/otlp.go:317]: I'm definitely ok with updating this to return an error as part of this PR.

**Review by TylerHelmuth — COMMENTED** [review]
@tvaintrob please add some test cases that hit the new error scenarios and a test for the new UnmarshalText function.

**Comment by tvaintrob** [comment]
@TylerHelmuth @songy23 updated the PR, i rather take care of the extra error cases in a different PR, feels out of scope for this one

**Review by songy23 — APPROVED** [review]
LGTM, please file an issue to return error on invalid content-type in requests.

**Comment by tvaintrob** [comment]
> LGTM, please file an issue to return error on invalid content-type in requests.

https://github.com/open-telemetry/opentelemetry-collector/issues/9413 created an issue for that, thanks

**Comment by tvaintrob** [comment]
@TylerHelmuth just waiting on your approval to merge this one

Thanks!

**Review by TylerHelmuth — COMMENTED** [review]
@tvaintrob could you add a could more test scenarios to cover the error cases `Codecov` is complaining about?

**Comment by tvaintrob** [comment]
> @tvaintrob could you add a could more test scenarios to cover the error cases `Codecov` is complaining about?

I had a talk with @songy23 about it, he said that it's ok to leave it like that, do you have any example on how to generate malformed data?

**Comment by TylerHelmuth** [comment]
I think you'd have to call the private functions directly. I am ok moving forward without these tests since they cover such unlikely, simple scenarios

**Review by sarathsp06 — APPROVED** [review]
LGTM

**Comment by mx-psi** [comment]
@tvaintrob please fix the merge conflict and I can merge afterwards :)

**Comment by tvaintrob** [comment]
@mx-psi Hi, resolved the merge conflict, thanks!

**Review by mx-psi — APPROVED** [review]

**Pull request merged**
